topostats/grainstats.py

```diff
@@ -0,0 +1,653 @@
+        nonzero_coordinates = edges.nonzero()
+        # Get vector representation of the points
+        # FIXME : Switched to list comprehension but should be unnecessary to create this as a list as we can use
+        # np.stack() to combine the arrays and use that...
+        # return np.stack(nonzero_coordinates, axis=1)
+        # edges = []
+        # for vector in np.transpose(nonzero_coordinates):
+        #     edges.append(list(vector))
+        # return edges
+        return [list(vector) for vector in np.transpose(nonzero_coordinates)]
+
+    def calculate_radius_stats(self, edges: list) -> Tuple:
+        """Class method that calculates the statistics relating to the radius. The radius in this context
+        is the distance from the centroid to points on the edge of the grain.
+
+        Parameters
+        ----------
+        edges: list
+            A 2D python list containing the coordinates of the edges of a grain.
+
+        Returns
+        -------
+        Tuple[float]
+            A tuple of the minimum, maximum, mean and median radius of the grain
+        """
+        # Convert the edges to the form of a numpy array
+        # edges = np.array(edges)
+        # Calculate the centroid of the grain
+        centroid = self._calculate_centroid(edges)
+        # Calculate the displacement
+        displacements = self._calculate_displacement(edges, centroid)
+        # Calculate the radius of each point
+        radii = self._calculate_radius(displacements)
+        return {"min": np.min(radii), "max": np.max(radii), "mean": np.mean(radii), "median": np.median(radii)}
+
+    def _calculate_centroid(self, edges: np.array) -> tuple:
+        """Calculate the centroid of a bounding box.
+
+        Parameters
+        ----------
+        edges: list
+            A 2D python list containing the co-ordinates of the edges of a grain.
+
+        Returns
+        -------
+        tuple
+            The co-ordinates of the centroid.
+        """
+        # FIXME : Remove once we have a numpy array returned by calculate_edges
+        edges = np.array(edges)
+        return (sum(edges[:, 0]) / len(edges), sum(edges[:, 1] / len(edges)))
+
+    def _calculate_displacement(self, edges: np.array, centroid: tuple) -> np.array:
+        """Calculate the displacement between the centroid and edges"""
+        # FIXME : Remove once we have a numpy array returned by calculate_edges
+        return np.array(edges) - centroid
+
+    def _calculate_radius(self, displacements) -> np.array:
+        """Calculate the radius of each point from the centroid
+
+        Parameters
+        ----------
+        displacements: List[list]
+
+        Retrurns
+        --------
+        np.array
+        """
+        return np.array([np.sqrt(radius[0] ** 2 + radius[1] ** 2) for radius in displacements])
+
+    def convex_hull(self, grain_mask: np.ndarray, edges: list, output_dir: Path, debug: bool = False):
```

Comment:
Is this unit tested anywhere?

Reply:
For this and `graham_scan()` functions I couldn't work out a sensible test and think both are candidates for refactoring into smaller methods (some of which I've done and in linting the code this morning pulled out another section of code). I was wary of going too far though as it was known there would be changes.

I'll create an Issue for outstanding code that needs unit-tests.

Reply:
Good plan.